File: MyAFBase/Shared/WidgetDataStore.swift
import Foundation

enum WidgetDataStore {
    static let appGroupID = "group.com.ryanladuca.MyAFBase"

    private static let readinessKey = "readinessWidgetSnapshot"
    private static let weatherKey = "weatherWidgetSnapshot"
    private static let openNowKey = "openNowWidgetSnapshot"
    private static let emergencyKey = "emergencyWidgetSnapshot"
    private static let payKey = "payWidgetSnapshot"
    private static let warTrackerKey = "warTrackerWidgetSnapshot"

    static func saveReadiness(_ snapshot: ReadinessWidgetSnapshot) {
        save(snapshot, forKey: readinessKey)
    }

    static func loadReadiness() -> ReadinessWidgetSnapshot? {
        load(ReadinessWidgetSnapshot.self, forKey: readinessKey)
    }

    static func saveWeather(_ snapshot: WeatherWidgetSnapshot) {
        save(snapshot, forKey: weatherKey)
    }

    static func loadWeather() -> WeatherWidgetSnapshot? {
        load(WeatherWidgetSnapshot.self, forKey: weatherKey)
    }

    static func saveOpenNow(_ snapshot: OpenNowWidgetSnapshot) {
        save(snapshot, forKey: openNowKey)
    }

    static func loadOpenNow() -> OpenNowWidgetSnapshot? {
        load(OpenNowWidgetSnapshot.self, forKey: openNowKey)
    }

    static func saveEmergency(_ snapshot: EmergencyWidgetSnapshot) {
        save(snapshot, forKey: emergencyKey)
    }

    static func loadEmergency() -> EmergencyWidgetSnapshot? {
        load(EmergencyWidgetSnapshot.self, forKey: emergencyKey)
    }

    static func savePay(_ snapshot: PayWidgetSnapshot) {
        save(snapshot, forKey: payKey)
    }

    static func loadPay() -> PayWidgetSnapshot? {
        load(PayWidgetSnapshot.self, forKey: payKey)
    }

    static func saveWARTracker(_ snapshot: WARWidgetSnapshot) {
        save(snapshot, forKey: warTrackerKey)
    }

    static func loadWARTracker() -> WARWidgetSnapshot? {
        load(WARWidgetSnapshot.self, forKey: warTrackerKey)
    }

    private static func save<T: Encodable>(_ value: T, forKey key: String) {
        guard let defaults = UserDefaults(suiteName: appGroupID),
              let data = try? JSONEncoder().encode(value) else {
            return
        }
        defaults.set(data, forKey: key)
    }

    private static func load<T: Decodable>(_ type: T.Type, forKey key: String) -> T? {
        guard let defaults = UserDefaults(suiteName: appGroupID),
              let data = defaults.data(forKey: key) else {
            return nil
        }
        return try? JSONDecoder().decode(type, from: data)
    }
}

File: android/app/src/main/java/com/ryanladuca/myafbase/widget/WidgetDataStore.kt
package com.ryanladuca.myafbase.widget

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val Context.widgetDataStore: DataStore<Preferences> by preferencesDataStore(
    name = "widget_snapshots",
)

class WidgetDataStore(
    private val context: Context,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val readinessKey = stringPreferencesKey("readiness")
    private val weatherKey = stringPreferencesKey("weather")
    private val openNowKey = stringPreferencesKey("open_now")
    private val emergencyKey = stringPreferencesKey("emergency")
    private val payKey = stringPreferencesKey("pay")
    private val warTrackerKey = stringPreferencesKey("war_tracker")

    suspend fun saveReadiness(snapshot: ReadinessWidgetSnapshot) = save(readinessKey, snapshot)
    suspend fun saveWeather(snapshot: WeatherWidgetSnapshot) = save(weatherKey, snapshot)
    suspend fun saveOpenNow(snapshot: OpenNowWidgetSnapshot) = save(openNowKey, snapshot)
    suspend fun saveEmergency(snapshot: EmergencyWidgetSnapshot) = save(emergencyKey, snapshot)
    suspend fun savePay(snapshot: PayWidgetSnapshot) = save(payKey, snapshot)
    suspend fun saveWarTracker(snapshot: WarWidgetSnapshot) = save(warTrackerKey, snapshot)

    suspend fun loadReadiness(): ReadinessWidgetSnapshot? = load(readinessKey)
    suspend fun loadWeather(): WeatherWidgetSnapshot? = load(weatherKey)
    suspend fun loadOpenNow(): OpenNowWidgetSnapshot? = load(openNowKey)
    suspend fun loadEmergency(): EmergencyWidgetSnapshot? = load(emergencyKey)
    suspend fun loadPay(): PayWidgetSnapshot? = load(payKey)
    suspend fun loadWarTracker(): WarWidgetSnapshot? = load(warTrackerKey)

    fun observePay() = context.widgetDataStore.data.map { prefs ->
        prefs[payKey]?.let { decode<PayWidgetSnapshot>(it) }
    }

    private suspend inline fun <reified T> save(key: Preferences.Key<String>, value: T) {
        context.widgetDataStore.edit { prefs ->
            prefs[key] = json.encodeToString(value)
        }
    }

    private suspend inline fun <reified T> load(key: Preferences.Key<String>): T? {
        val raw = context.widgetDataStore.data.first()[key] ?: return null
        return decode(raw)
    }

    private inline fun <reified T> decode(raw: String): T? =
        runCatching { json.decodeFromString<T>(raw) }.getOrNull()
}
